File: src/test/java/dev/jiaming/ai_interview/common/RedisRequestGuardTests.java
package dev.jiaming.ai_interview.common;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.json.JsonMapper;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.mockito.junit.jupiter.MockitoSettings;
import org.mockito.quality.Strictness;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.ValueOperations;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;
import org.springframework.web.server.ResponseStatusException;

@ExtendWith(MockitoExtension.class)
@MockitoSettings(strictness = Strictness.LENIENT)
class RedisRequestGuardTests {

	private final Map<String, String> redis = new ConcurrentHashMap<>();

	@Mock
	private StringRedisTemplate redisTemplate;

	@Mock
	private ValueOperations<String, String> valueOperations;

	private RedisRequestGuard guard;

	@BeforeEach
	void setUp() {
		when(redisTemplate.opsForValue()).thenReturn(valueOperations);
		when(valueOperations.get(anyString())).thenAnswer(invocation -> redis.get(invocation.getArgument(0)));
		when(valueOperations.setIfAbsent(anyString(), anyString(), any(Duration.class)))
			.thenAnswer(invocation -> redis.putIfAbsent(invocation.getArgument(0), invocation.getArgument(1)) == null);
		doAnswer(invocation -> {
			redis.put(invocation.getArgument(0), invocation.getArgument(1));
			return null;
		}).when(valueOperations).set(anyString(), anyString(), any(Duration.class));
		when(valueOperations.increment(anyString())).thenAnswer(invocation -> {
			String key = invocation.getArgument(0);
			long nextValue = Long.parseLong(redis.getOrDefault(key, "0")) + 1;
			redis.put(key, Long.toString(nextValue));
			return nextValue;
		});
		when(redisTemplate.expire(anyString(), any(Duration.class))).thenReturn(true);

		guard = new RedisRequestGuard(
			redisTemplate,
			new RedisUsageProperties(
				new RedisUsageProperties.RateLimit(true, 60, 2, 2),
				new RedisUsageProperties.Idempotency(true, 86_400)
			),
			objectMapper()
		);
	}

	@AfterEach
	void tearDown() {
		RequestContextHolder.resetRequestAttributes();
	}

	@Test
	void replaysCachedResponseForSameIdempotencyKeyAndPayload() {
		requestWithIdempotencyKey("retry-key");
		AtomicInteger calls = new AtomicInteger();

		CachedResponse first = guard.withIdempotentRetryCache(
			"assessment",
			List.of("resume", "job"),
			CachedResponse.class,
			() -> new CachedResponse("run-" + calls.incrementAndGet())
		);
		CachedResponse second = guard.withIdempotentRetryCache(
			"assessment",
			List.of("resume", "job"),
			CachedResponse.class,
			() -> new CachedResponse("run-" + calls.incrementAndGet())
		);

		assertThat(first).isEqualTo(new CachedResponse("run-1"));
		assertThat(second).isEqualTo(first);
		assertThat(calls).hasValue(1);
	}

	@Test
	void rejectsSameIdempotencyKeyForDifferentPayload() {
		requestWithIdempotencyKey("retry-key");

		guard.withIdempotentRetryCache(
			"assessment",
			List.of("resume-a"),
			CachedResponse.class,
			() -> new CachedResponse("saved")
		);

		assertThatThrownBy(() -> guard.withIdempotentRetryCache(
			"assessment",
			List.of("resume-b"),
			CachedResponse.class,
			() -> new CachedResponse("should-not-run")
		))
			.isInstanceOf(ResponseStatusException.class)
			.hasMessageContaining("409 CONFLICT")
			.hasMessageContaining("Idempotency-Key");
	}

	@Test
	void bypassesIdempotencyCacheWhenHeaderIsMissing() {
		requestWithIdempotencyKey(null);
		AtomicInteger calls = new AtomicInteger();

		CachedResponse first = guard.withIdempotentRetryCache(
			"assessment",
			List.of("resume"),
			CachedResponse.class,
			() -> new CachedResponse("run-" + calls.incrementAndGet())
		);
		CachedResponse second = guard.withIdempotentRetryCache(
			"assessment",
			List.of("resume"),
			CachedResponse.class,
			() -> new CachedResponse("run-" + calls.incrementAndGet())
		);

		assertThat(first).isEqualTo(new CachedResponse("run-1"));
		assertThat(second).isEqualTo(new CachedResponse("run-2"));
		assertThat(calls).hasValue(2);
	}

	@Test
	void rateLimitsAiRequestsPerClientBucket() {
		requestWithIdempotencyKey(null);

		guard.assertAiAllowed("assessment");
		guard.assertAiAllowed("assessment");

		assertThatThrownBy(() -> guard.assertAiAllowed("assessment"))
			.isInstanceOf(ResponseStatusException.class)
			.hasMessageContaining("429 TOO_MANY_REQUESTS")
			.hasMessageContaining("Too many assessment requests");
	}

	@Test
	void rateLimitsUploadRequestsSeparatelyFromAiRequests() {
		requestWithIdempotencyKey(null);

		guard.assertUploadAllowed();
		guard.assertUploadAllowed();

		assertThatThrownBy(guard::assertUploadAllowed)
			.isInstanceOf(ResponseStatusException.class)
			.hasMessageContaining("429 TOO_MANY_REQUESTS")
			.hasMessageContaining("resume-upload");
	}

	@Test
	void namespacesRateLimitAndIdempotencyKeys() {
		requestWithIdempotencyKey("retry-key");
		guard.assertAiAllowed("assessment");
		guard.withIdempotentRetryCache(
			"assessment", List.of("resume"), CachedResponse.class, () -> new CachedResponse("saved")
		);

		assertThat(redis.keySet()).allMatch(key -> key.startsWith("ai-interview:v3:"));
	}

	@Test
	void doesNotTrustClientSuppliedForwardedAddressForRateLimits() {
		MockHttpServletRequest first = new MockHttpServletRequest();
		first.setRemoteAddr("203.0.113.10");
		first.addHeader("X-Forwarded-For", "198.51.100.1");
		RequestContextHolder.setRequestAttributes(new ServletRequestAttributes(first));
		guard.assertAiAllowed("assessment");

		MockHttpServletRequest second = new MockHttpServletRequest();
		second.setRemoteAddr("203.0.113.10");
		second.addHeader("X-Forwarded-For", "198.51.100.2");
		RequestContextHolder.setRequestAttributes(new ServletRequestAttributes(second));
		guard.assertAiAllowed("assessment");

		assertThatThrownBy(() -> guard.assertAiAllowed("assessment"))
			.isInstanceOf(ResponseStatusException.class)
			.hasMessageContaining("429 TOO_MANY_REQUESTS");
	}

	@Test
	void doesNotTouchRedisWhenIdempotencyIsDisabled() {
		requestWithIdempotencyKey("retry-key");
		RedisRequestGuard disabledGuard = new RedisRequestGuard(
			redisTemplate,
			new RedisUsageProperties(
				new RedisUsageProperties.RateLimit(true, 60, 2, 2),
				new RedisUsageProperties.Idempotency(false, 86_400)
			),
			objectMapper()
		);

		CachedResponse response = disabledGuard.withIdempotentRetryCache(
			"assessment",
			List.of("resume"),
			CachedResponse.class,
			() -> new CachedResponse("uncached")
		);

		assertThat(response).isEqualTo(new CachedResponse("uncached"));
		verifyNoInteractions(valueOperations);
	}

	private void requestWithIdempotencyKey(String idempotencyKey) {
		MockHttpServletRequest request = new MockHttpServletRequest();
		request.setRemoteAddr("203.0.113.10");
		if (idempotencyKey != null) {
			request.addHeader("Idempotency-Key", idempotencyKey);
		}
		RequestContextHolder.setRequestAttributes(new ServletRequestAttributes(request));
	}

	private ObjectMapper objectMapper() {
		return JsonMapper.builder()
			.findAndAddModules()
			.build();
	}

	private record CachedResponse(String value) {
	}
}

File: src/test/kotlin/dev/jiaming/ai_interview/common/RedisRequestGuardTests.kt
package dev.jiaming.ai_interview.common

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.json.JsonMapper
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.quality.Strictness
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.data.redis.core.ValueOperations
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import org.springframework.web.server.ResponseStatusException
import java.time.Duration
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

@ExtendWith(MockitoExtension::class)
@MockitoSettings(strictness = Strictness.LENIENT)
class RedisRequestGuardTests {

    private val redis = ConcurrentHashMap<String, String>()

    @Mock
    private lateinit var redisTemplate: StringRedisTemplate

    @Mock
    private lateinit var valueOperations: ValueOperations<String, String>

    private lateinit var guard: RedisRequestGuard

    @BeforeEach
    fun setUp() {
        Mockito.`when`(redisTemplate.opsForValue()).thenReturn(valueOperations)
        Mockito.`when`(valueOperations.get(anyString())).thenAnswer { redis[it.getArgument(0)] }
        Mockito.`when`(valueOperations.setIfAbsent(anyString(), anyString(), any(Duration::class.java)))
            .thenAnswer { redis.putIfAbsent(it.getArgument(0), it.getArgument(1)) == null }
        Mockito.doAnswer {
            redis[it.getArgument(0)] = it.getArgument(1)
            null
        }.`when`(valueOperations).set(anyString(), anyString(), any(Duration::class.java))
        Mockito.`when`(valueOperations.increment(anyString())).thenAnswer {
            val key = it.getArgument<String>(0)
            val nextValue = redis.getOrDefault(key, "0").toLong() + 1
            redis[key] = nextValue.toString()
            nextValue
        }
        Mockito.`when`(redisTemplate.expire(anyString(), any(Duration::class.java))).thenReturn(true)
        guard = RedisRequestGuard(
            redisTemplate,
            RedisUsageProperties(
                RedisUsageProperties.RateLimit(true, 60, 2, 2),
                RedisUsageProperties.Idempotency(true, 86_400)
            ),
            objectMapper()
        )
    }

    @AfterEach
    fun tearDown() = RequestContextHolder.resetRequestAttributes()

    @Test
    fun replaysCachedResponseForSameIdempotencyKeyAndPayload() {
        requestWithIdempotencyKey("retry-key")
        val calls = AtomicInteger()
        val first = guard.withIdempotentRetryCache("assessment", listOf("resume", "job"), CachedResponse::class.java) {
            CachedResponse("run-${calls.incrementAndGet()}")
        }
        val second = guard.withIdempotentRetryCache("assessment", listOf("resume", "job"), CachedResponse::class.java) {
            CachedResponse("run-${calls.incrementAndGet()}")
        }

        assertThat(first).isEqualTo(CachedResponse("run-1"))
        assertThat(second).isEqualTo(first)
        assertThat(calls).hasValue(1)
    }

    @Test
    fun rejectsSameIdempotencyKeyForDifferentPayload() {
        requestWithIdempotencyKey("retry-key")
        guard.withIdempotentRetryCache("assessment", listOf("resume-a"), CachedResponse::class.java) { CachedResponse("saved") }

        assertThatThrownBy {
            guard.withIdempotentRetryCache("assessment", listOf("resume-b"), CachedResponse::class.java) {
                CachedResponse("should-not-run")
            }
        }.isInstanceOf(ResponseStatusException::class.java)
            .hasMessageContaining("409 CONFLICT")
            .hasMessageContaining("Idempotency-Key")
    }

    @Test
    fun bypassesIdempotencyCacheWhenHeaderIsMissing() {
        requestWithIdempotencyKey(null)
        val calls = AtomicInteger()
        val first = guard.withIdempotentRetryCache("assessment", listOf("resume"), CachedResponse::class.java) {
            CachedResponse("run-${calls.incrementAndGet()}")
        }
        val second = guard.withIdempotentRetryCache("assessment", listOf("resume"), CachedResponse::class.java) {
            CachedResponse("run-${calls.incrementAndGet()}")
        }

        assertThat(first).isEqualTo(CachedResponse("run-1"))
        assertThat(second).isEqualTo(CachedResponse("run-2"))
        assertThat(calls).hasValue(2)
    }

    @Test
    fun rateLimitsAiRequestsPerClientBucket() {
        requestWithIdempotencyKey(null)
        guard.assertAiAllowed("assessment")
        guard.assertAiAllowed("assessment")

        assertThatThrownBy { guard.assertAiAllowed("assessment") }
            .isInstanceOf(ResponseStatusException::class.java)
            .hasMessageContaining("429 TOO_MANY_REQUESTS")
            .hasMessageContaining("Too many assessment requests")
    }

    @Test
    fun rateLimitsUploadRequestsSeparatelyFromAiRequests() {
        requestWithIdempotencyKey(null)
        guard.assertUploadAllowed()
        guard.assertUploadAllowed()

        assertThatThrownBy { guard.assertUploadAllowed() }
            .isInstanceOf(ResponseStatusException::class.java)
            .hasMessageContaining("429 TOO_MANY_REQUESTS")
            .hasMessageContaining("resume-upload")
    }

    @Test
    fun namespacesRateLimitAndIdempotencyKeys() {
        requestWithIdempotencyKey("retry-key")
        guard.assertAiAllowed("assessment")
        guard.withIdempotentRetryCache("assessment", listOf("resume"), CachedResponse::class.java) { CachedResponse("saved") }

        assertThat(redis.keys).allMatch { it.startsWith("ai-interview:v3:") }
    }

    @Test
    fun doesNotTrustClientSuppliedForwardedAddressForRateLimits() {
        val first = MockHttpServletRequest().apply {
            remoteAddr = "203.0.113.10"
            addHeader("X-Forwarded-For", "198.51.100.1")
        }
        RequestContextHolder.setRequestAttributes(ServletRequestAttributes(first))
        guard.assertAiAllowed("assessment")
        val second = MockHttpServletRequest().apply {
            remoteAddr = "203.0.113.10"
            addHeader("X-Forwarded-For", "198.51.100.2")
        }
        RequestContextHolder.setRequestAttributes(ServletRequestAttributes(second))
        guard.assertAiAllowed("assessment")

        assertThatThrownBy { guard.assertAiAllowed("assessment") }
            .isInstanceOf(ResponseStatusException::class.java)
            .hasMessageContaining("429 TOO_MANY_REQUESTS")
    }

    @Test
    fun doesNotTouchRedisWhenIdempotencyIsDisabled() {
        requestWithIdempotencyKey("retry-key")
        val disabledGuard = RedisRequestGuard(
            redisTemplate,
            RedisUsageProperties(
                RedisUsageProperties.RateLimit(true, 60, 2, 2),
                RedisUsageProperties.Idempotency(false, 86_400)
            ),
            objectMapper()
        )

        val response = disabledGuard.withIdempotentRetryCache(
            "assessment", listOf("resume"), CachedResponse::class.java
        ) { CachedResponse("uncached") }

        assertThat(response).isEqualTo(CachedResponse("uncached"))
        Mockito.verifyNoInteractions(valueOperations)
    }

    private fun requestWithIdempotencyKey(idempotencyKey: String?) {
        val request = MockHttpServletRequest().apply {
            remoteAddr = "203.0.113.10"
            if (idempotencyKey != null) addHeader("Idempotency-Key", idempotencyKey)
        }
        RequestContextHolder.setRequestAttributes(ServletRequestAttributes(request))
    }

    private fun objectMapper(): ObjectMapper = JsonMapper.builder().findAndAddModules().build()

    data class CachedResponse @JsonCreator constructor(@param:JsonProperty("value") val value: String)
}
